### PR TITLE
Add podman machine test suite

### PR DESCRIPTION
**Pull request by baude**
This PR introduces a test suite for podman machine.  It can currently be
run on developers' local machines and is not part of the official CI
testing; however, the expectation is that any work on machine should
come with an accompanying test.

At present, the test must be run on Linux.  It is untested on Darwin.
There is no Makefile target for the test.  It can be run like `ginkgo -v
pkg/machine/test/.`.  It should be run as a unprivileged user.

Signed-off-by: Brent Baude <bbaude@redhat.com>

### CODE REVIEW

File: pkg/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -1408,7 +1408,7 @@ const (
 	PodPending PodPhase = "Pending"
 	// PodRunning means the pod has been bound to a node and all of the containers have been started.
 	// At least one container is still running or is in the process of being restarted.
-	PodRunning PodPhase = "Running"
+	PodRunning PodPhase = "isRunning"
```
Comment (vrothberg) [inline pkg/k8s.io/api/core/v1/types.go:1411]: The `s/Running/isRunning/` changes look like a too aggressive search-and-replace.

File: Makefile [hunk 2]
```diff
@@ -56,7 +56,8 @@ BUILDTAGS ?= \
 	$(shell hack/systemd_tag.sh) \
 	$(shell hack/libsubid_tag.sh) \
 	exclude_graphdriver_devicemapper \
-	seccomp
+	seccomp \
+	native_rootless
```
Comment (vrothberg) [inline Makefile:60]: Is the `native_rootless` change still needed?  If so, could we move it into a dedicated commit (and probably invert the logic to not break downstream packaging).

Comment (Luap99) [reply Makefile:60]: Yes if needed it should be reverted, however for the purpose of machine test we could use the podman-remote binary which should never join the userns.

File: cmd/podman/machine/machine.go [hunk 3]
```diff
@@ -24,6 +24,10 @@ var (
 		PersistentPreRunE:  noOp,
 		PersistentPostRunE: noOp,
 		RunE:               validate.SubCommandExists,
+		Annotations: map[string]string{
+			registry.UnshareNSRequired: "",
+			registry.ParentNSRequired:  "",
+		},
```
Comment (Luap99) [inline cmd/podman/machine/machine.go:30]: Please remove this. This breaks the machine command.

File: hack/golangci-lint.sh [hunk 4]
```diff
@@ -10,9 +10,9 @@ BUILD_TAGS[abi]="${BUILD_TAGS[default]},!remoteclient"
 BUILD_TAGS[tunnel]="${BUILD_TAGS[default]},remote,remoteclient"
 
 declare -A SKIP_DIRS
-SKIP_DIRS[abi]=""
+SKIP_DIRS[abi]="pkg/machine/e2e"
```
Comment (Luap99) [inline hack/golangci-lint.sh:13]: Why skip this, linting is super useful. I am against excluding new code. You should rather fix the reported problems.

Comment (baude) [reply hack/golangci-lint.sh:13]: im skipping this because not all the methods are being used yet.  It can be linted once that is the case.  And I didn't want to spread around a bunch of single nolints.

**Review by Luap99 — CHANGES_REQUESTED** [review]
still plenty of unrelated s/Running/isRunning/ changes, please revert them all

**Comment by TomSweeneyRedHat** [comment]
LGTM
@edsantiago PTAL

**Review by vrothberg — COMMENTED** [review]
Nit: the changes in `pkg/rootless` could be reverted.  Other than that, LGTM

File: pkg/machine/qemu/machine.go [hunk 5]
```diff
@@ -517,14 +514,22 @@ func (v *MachineVM) Start(name string, _ machine.StartOptions) error {
 	if err != nil {
 		return err
 	}
-
 	fd, err := qemuSocketConn.(*net.UnixConn).File()
 	if err != nil {
 		return err
 	}
 
+	dnr, err := os.Open("/dev/null")
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:522]: Os.Open() only opens with read access so you cannot use it for stdout/err file descriptors.

File: pkg/machine/qemu/machine.go [hunk 6]
```diff
@@ -517,14 +514,22 @@ func (v *MachineVM) Start(name string, _ machine.StartOptions) error {
...
+		return err
+	}
+	dnw, err := os.Open("/dev/null")
+	if err != nil {
+		return err
+	}
+
 	attr := new(os.ProcAttr)
-	files := []*os.File{os.Stdin, os.Stdout, os.Stderr, fd}
+	files := []*os.File{dnw, dnr, dnr, fd}
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:532]: Should we make this conditional on debug level? So when debug we still attach stdout/err to get the qemu output. Otherwise the statement on line 540 is pretty useless.

File: pkg/machine/qemu/machine.go [hunk 7]
```diff
@@ -579,7 +584,10 @@ func (v *MachineVM) Start(name string, _ machine.StartOptions) error {
 	if err != nil {
 		return err
 	}
-
+	conn.Close()
+	dnr.Close()
+	dnw.Close()
+	fd.Close()
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:590]: These should be `defer fd.Close()` after they are opened. Right now this code will leak the open fds when an error occurs and it returns early.

File: pkg/machine/qemu/machine.go [hunk 8]
```diff
@@ -1085,9 +1093,6 @@ func (v *MachineVM) startHostNetworking() (string, apiForwardingState, error) {
 	}
 
 	attr := new(os.ProcAttr)
-	// Pass on stdin, stdout, stderr
-	files := []*os.File{os.Stdin, os.Stdout, os.Stderr}
-	attr.Files = files
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:1090]: Will this attach any fds to stdin/out/err for gvproxy?  I am not sure why you set /dev/null for qemu but not here for gvproxy?

File: pkg/rootless/rootless_linux.c [hunk 9]
```diff
@@ -215,7 +215,7 @@ can_use_shortcut ()
   bool ret = true;
   int argc;
 
-#ifdef DISABLE_JOIN_SHORTCUT
+#ifndef ENABLE_JOIN_SHORTCUT
```
Comment (Luap99) [inline pkg/rootless/rootless_linux.c:218]: please revert this change since it is not needed

File: pkg/rootless/rootless_linux.go [hunk 10]
```diff
@@ -30,7 +30,7 @@ import (
 )
 
 /*
-#cgo remoteclient CFLAGS: -Wall -Werror -DDISABLE_JOIN_SHORTCUT
+#cgo native_rootless CFLAGS: -Wall -Werror -DDISABLE_JOIN_SHORTCUT
```
Comment (Luap99) [inline pkg/rootless/rootless_linux.go:33]: please revert this change

File: test/e2e/prune_test.go [hunk 11]
```diff
@@ -359,7 +359,7 @@ var _ = Describe("Podman prune", func() {
 		// One Pod should not be pruned as it was running
 		Expect(podmanTest.NumberOfPods()).To(Equal(1))
 
-		// Running pods infra container should not be pruned.
+		// isRunning pods infra container should not be pruned.
```
Comment (Luap99) [inline test/e2e/prune_test.go:362]: revert this change

File: pkg/machine/qemu/machine.go (hunk 5 above)
Comment (mheon) [reply pkg/machine/qemu/machine.go:522]: @baude You want https://pkg.go.dev/os#OpenFile so you can specify how the file will be opened

File: pkg/machine/qemu/machine.go (hunk 6 above)
Comment (baude) [reply pkg/machine/qemu/machine.go:532]: All of the output from qemu in this case ends up in its display (the qemu window), so nothing is actually lost.

File: pkg/machine/qemu/machine.go (hunk 8 above)
Comment (baude) [reply pkg/machine/qemu/machine.go:1090]: the reason behind this, right or wrong, was because for we do actually pass in an fd for socket use.  To keep the order of the fd the same, i passed in /dev/nulls.  And then with gvproxy, I just removed the fds because we really don't need them.  Would you rather I went through the exercise of not passing in /dev/nulls and fix the fd ordering for the socket?

File: pkg/machine/qemu/machine.go [hunk 12]
```diff
@@ -572,14 +580,14 @@ func (v *MachineVM) Start(name string, _ machine.StartOptions) error {
 		time.Sleep(wait)
 		wait++
 	}
+	defer conn.Close()
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:583]: This must be done after the error check below, otherwise it causes a panic on error

File: pkg/machine/qemu/machine.go (hunk 8 above)
Comment (Luap99) [reply pkg/machine/qemu/machine.go:1090]: The problem is you cannot just not set fds for the child process. If there are no fds and the child starts opening fds then they will get placed on 0,1 and 2 which causes all kinds of problems. For example if gvproxy tries to write something to stdout/err now it will write to a random fd and this will likely fail or cause other side effects.

File: pkg/machine/qemu/machine.go (hunk 6 above)
Comment (Luap99) [reply pkg/machine/qemu/machine.go:535]: The fds are inverted. stdin should be dnr, and stdou/err should be dnw.

File: pkg/machine/qemu/machine.go [hunk 13]
```diff
@@ -517,14 +514,25 @@ func (v *MachineVM) Start(name string, _ machine.StartOptions) error {
...
+	dnr, err := os.OpenFile("/dev/null", os.O_RDONLY, 0755)
+	if err != nil {
+		return err
+	}
+	dnw, err := os.OpenFile("/dev/null", os.O_WRONLY, 0755)
+	if err != nil {
+		return err
+	}
+
+	defer dnr.Close()
```
Comment (Luap99) [inline pkg/machine/qemu/machine.go:531]: the defer statement should be after the file is opened after line 525. If the second open file fails it will leak this fd otherwise.
same for the qemuSocketConn fd.

**Review by Luap99 — COMMENTED** [review]
LGTM

**Review by Luap99 — APPROVED** [review]
/lgtm
/hold
